packages/pwa-kit-dev/src/utils/script-utils.ts

```diff
@@ -0,0 +1,312 @@
+            credentials: params.credentials
+        }
+    }
+
+    private getAuthHeader(): StringMap {
+        const {username, api_key} = this.opts.credentials
+        const encoded = Buffer.from(`${username}:${api_key}`, 'binary').toString('base64')
+        return {Authorization: `Basic ${encoded}`}
+    }
+
+    private async getHeaders(extras: StringMap): Promise<StringMap> {
+        const pkg = await getPkgJSON()
+        return {
+            'User-Agent': `${pkg.name}@${pkg.version}`,
+            ...this.getAuthHeader(),
+            ...extras
+        }
+    }
+
+    private async throwForStatus(res) {
+        if (res.status < 400) {
+            return
+        }
+
+        let error
+        try {
+            error = await res.json()
+        } catch (err) {
+            error = {} // Cloud doesn't always return JSON
+        }
+
+        throw new Error(
+            [
+                `HTTP ${res.status}`,
+                error.message || res.body,
+                `For more information visit ${error.docs_url || DEFAULT_DOCS_URL}`
+            ].join('\n')
+        )
+    }
+
+    async push(bundle: Bundle, projectSlug: string, target?: string) {
+        const base = `api/projects/${projectSlug}/builds/`
+        const pathname = target ? base + `${target}/` : base
+        const url = new URL(this.opts.origin)
+        url.pathname = pathname
+
+        const body = Buffer.from(JSON.stringify(bundle))
+        const headers = await this.getHeaders({'Content-Length': body.length.toString()})
+
+        const res = await this.opts.fetch(url.toString(), {
+            body,
+            method: 'POST',
+            headers
+        })
+        await this.throwForStatus(res)
+        return await res.json()
+    }
+
+    async createLoggingToken(project: string, environment: string): Promise<string> {
+        const url = new URL(this.opts.origin)
+        url.pathname = `/api/projects/${project}/target/${environment}/jwt/`
+        const headers = await this.getHeaders({
+            Authorization: `Bearer ${this.opts.credentials.api_key}`
+        })
+        const res = await this.opts.fetch(url.toString(), {
+            method: 'POST',
+            headers
+        })
+        await this.throwForStatus(res)
+        const data = await res.json()
+        return data['token']
+    }
+}
+
+export const defaultMessage = (gitInstance: git = git): string => {
+    try {
+        return `${gitInstance.branch()}: ${gitInstance.short()}`
+    } catch (err) {
+        if (err.code === 'ENOENT') {
+            console.log(
+                'Using default bundle message as no message was provided and not in a Git repo.'
+            )
+        }
+        return 'PWA Kit Bundle'
+    }
+}
+
+interface CreateBundleArgs {
+    message: string | null | undefined
+    ssr_parameters: any
+    ssr_only: string[]
+    ssr_shared: string[]
+    buildDirectory: string
+    projectSlug: string
+}
+
+export const createBundle = async ({
```

Comment:
Because this code is ancient and it's always been the case. I don't want to touch the archiver stuff, which has been preserved from before.

This feedback is legit – I don't want to fix **everything** about the upload code in this one PR.

These bundles are 100's of MB in size. Taking a step back, if I asked you _"should we do this in memory or on disk?"_, I think you would want to keep them on disk. It's the in-memory part that looks weird, I think.

Reply:
It's a significant improvement that the updated code doesn't just dump `build.tar` into your project and leave it there for no reason!